consensus/spos/commonSubround/subroundStartRound.go

```diff
@@ -255,3 +270,42 @@ func (sr *SubroundStartRound) generateNextConsensusGroup(roundIndex int64) error
 
 	return nil
 }
+
+// EpochStartPrepare wis called when an epoch start event is observed, but not yet confirmed/committed.
+// Some components may need to do initialisation on this event
+func (sr *SubroundStartRound) EpochStartPrepare(metaHeader data.HeaderHandler) {
+	log.Trace(fmt.Sprintf("epoch %d start prepare in consensus", metaHeader.GetEpoch()))
+}
+
+// EpochStartAction is called upon a start of epoch event.
+func (sr *SubroundStartRound) EpochStartAction(hdr data.HeaderHandler) {
+	log.Trace(fmt.Sprintf("epoch %d start action in consensus", hdr.GetEpoch()))
+
+	sr.changeEpoch(hdr)
+}
+
+func (sr *SubroundStartRound) changeEpoch(header data.HeaderHandler) {
+	publicKeysPrevEpoch, err := sr.NodesCoordinator().GetAllValidatorsPublicKeys(header.GetEpoch() - 1)
+	if err != nil {
+		log.Error(fmt.Sprintf("epoch %d: %s", header.GetEpoch()-1, err.Error()))
+		return
+	}
+
+	publicKeysNewEpoch, err := sr.NodesCoordinator().GetAllValidatorsPublicKeys(header.GetEpoch())
+	if err != nil {
+		log.Error(fmt.Sprintf("epoch %d: %s", header.GetEpoch(), err.Error()))
+		return
+	}
+
+	shardEligible := make(map[string]struct{}, 0)
+	shardId := sr.ShardCoordinator().SelfId()
+
+	for _, pubKey := range publicKeysPrevEpoch[shardId] {
+		shardEligible[string(pubKey)] = struct{}{}
+	}
+	for _, pubKey := range publicKeysNewEpoch[shardId] {
+		shardEligible[string(pubKey)] = struct{}{}
+	}
+
+	sr.SetEligibleList(shardEligible)
```

Comment:
So the `roundConsensus.eligibleNodes` map will contain the keys of both the validators from the _previous_ epoch and the _new_ one. Does this mean that if I call `roundConsensus.IsNodeInEligibleList()` and give it the public key of a Node that was eligible yesterday, but today the Node is in waiting, the method will tell me "yes, it's eligible"? Is this true? If yes, is it intentional?

Reply:
This is only used as an anti-flood mechanism, where messages on the consensus topic (of a shard) are accepted only if they come from one of the eligible nodes. Here we allow both previous epoch nodes and new epoch nodes to send messages on the consensus topic.

The reason for this is because in shards, the entry in the new epoch can be done with slight delay on the shards and not in the same round, while the majority of the nodes did not receive the trigger for the new epoch (latency), which would cause some rounds with no blocks.

Reply:
I'd argue that the anti-flood mechanism should explicitly ask "was this Node eligible yesterday?" using a separate function, rather than having `IsNodeInEligibleList()` return `true` for both today's and yesterday's Nodes. The name of this function definitely does not imply that, and anti-flooding should not rely on this implicit behavior.